compiler/src/lnntoamm/Types.ts

```diff
@@ -123,127 +193,279 @@ export default abstract class Type implements Equalable {
+    let isNothingNew = true;
+    genKeys.forEach((genName) => {
+      const thisGen = this.generics[genName];
+      let tyVal: Type;
+      if (thisGen === null) {
+        tyVal = Type.generate();
+      } else {
+        const duped = thisGen.dup();
+        if (duped === null) {
+          tyVal = thisGen;
+        } else {
+          tyVal = duped;
+          isNothingNew = false;
+        }
       }
-      return that.tempDelegate !== null && this.eq(that.tempDelegate);
-    } else {
+      duped.generics[genName] = tyVal;
+    });
+    return duped;
+  }
+
+  eq(that: Equalable): boolean {
+    if (!(that instanceof Opaque) || this.name !== that.name) {
       return false;
     }
+    const thisGens = Object.values(this.generics);
+    const thatGens = Object.values(that.generics);
+    return (
+      thisGens.length === thatGens.length &&
+      thisGens.every((thisGen, ii) => {
+        const thatGen = thatGens[ii];
+        if (thisGen === null || thatGen === null) {
+          return thisGen === thatGen;
+        } else {
+          return thisGen.eq(thatGen);
+        }
+      })
+    );
   }
 
-  fieldIndices(): FieldIndices {
-    return TODO("determine if it's even worth keeping the Builtin class");
+  fnselectOptions(): Type[] {
+    const genOptions = Object.values(this.generics).map((g) =>
+      g === null ? [g] : g.fnselectOptions(),
+    );
+    const opts = new Array<Type>();
+    const getIndices = matrixIndices(genOptions);
+    const toSolidify = new Opaque(this.name, Object.keys(this.generics));
+    for (
+      let indicesRes = getIndices.next();
+      !indicesRes.done;
+      indicesRes = getIndices.next()
+    ) {
+      const indices = indicesRes.value as number[];
+      opts.push(
+        toSolidify.solidify(
+          indices.map((optIdx, tyVarIdx) => genOptions[tyVarIdx][optIdx]),
+        ),
+      );
+    }
+    return opts;
   }
 
-  instance(): Type {
-    return this;
+  instance(opts?: InstanceOpts): Type {
+    const genNames = Object.keys(this.generics);
+    if (genNames.length === 0) {
+      // minor optimization: if there's no generics then we keep the same JS
+      // object to reduce the number of allocs
+      return this;
+    }
+    const instance = new Opaque(this.name, genNames);
+    for (const name of genNames) {
+      const thisGen = this.generics[name];
+      if (thisGen === null) {
+        throw new Error(
+          `Cannot get an instance of a generic Opaque type that hasn't been solidified`,
+        );
+      }
+      instance.generics[name] = thisGen.instance(opts);
+    }
+    return instance;
   }
 
   isFixed(): boolean {
     switch (this.name) {
-      case 'int64':
-      case 'int32':
-      case 'int16':
-      case 'int8':
-      case 'float64':
-      case 'float32':
-      case 'bool':
-      case 'void':
-        return true;
-      default:
+      case 'string':
         return false;
+      default:
+        return true;
```

Comment:
Can you revert this change? It's why `Result<int64>` got `reff` instead of `refv`.